llvm/lib/SYCLLowerIR/SYCLVirtualFunctionsAnalysis.cpp

```diff
@@ -109,8 +201,50 @@ SYCLVirtualFunctionsAnalysisPass::run(Module &M, ModuleAnalysisManager &MAM) {
   }
 
   // Emit a diagnostic if a kernel performs virtual function calls
-  for (const auto *K : Kernels) {
+  for (auto *K : KernelsToCheck)
     checkKernel(K, CallGraph);
+
+  // Cache to know which function is used by which kernels
+  FuncToFuncMapTy FunctionToKernels;
+  for (auto *K : AllKernels)
+    computeFunctionToKernelsMapping(K, CallGraph, FunctionToKernels);
+
+  for (const auto *F : IndirectlyCallableFuncs) {
+    StringRef Set = F->getFnAttribute("indirectly-callable").getValueAsString();
+
+    SmallVector<const GlobalVariable *, 4> VTables;
+    collectVTablesThatUseFunction(F, VTables);
+    SmallPtrSet<Function *, 8> KernelsToUpdate;
+
+    for (const auto *GV : VTables) {
+      // Find functions that use those vtables
+      SmallPtrSet<const Function *, 16> FunctionsThatUseVTables;
+      for (const auto *UU : GV->users())
+        collectEnclosingFunctions(UU, FunctionsThatUseVTables);
+      // And collect kernels that use those functions
+      for (const Function *FF : FunctionsThatUseVTables)
+        for (auto *K : FunctionToKernels[FF])
+          KernelsToUpdate.insert(K);
+    }
+
+    // Update or attach "calls-indirectly" attribute to those kernels
+    // indicating that they use virtual functions set 'Set'
+    for (Function *K : KernelsToUpdate) {
+      if (!K->hasFnAttribute("calls-indirectly"))
+        K->addFnAttr("calls-indirectly", Set);
+      else {
+        StringRef UsedSets =
+            K->getFnAttribute("calls-indirectly").getValueAsString();
+        if (UsedSets.contains(Set))
+          continue;
+
+        K->removeFnAttr("calls-indirectly");
+        SmallString<64> NewAttr = UsedSets;
+        NewAttr += ",";
+        NewAttr += Set;
+        K->addFnAttr("calls-indirectly", NewAttr.str());
```

Comment:
do we need to do any duplicate checking here? i'm not even sure if it's possible, but even if it is, it might not even matter, depending on how the runtime uses the value

Reply:
Ah right, thanks.